tests/test_files_cloud.py

```diff
@@ -0,0 +1,97 @@
+"""
+This module contains integration tests which verify the behavior of the storage
+module against supported cloud storage providers. The tests for the local
+(directory-based) storage backend are re-used and run for each supported cloud
+storage provider. To add tests for a new provider, create a new subclass of
+CloudTestFile.
+
+To run tests for Google Cloud Platform, set the following environment variables:
+- GCP_STORAGE_KEY
+- GCP_STORAGE_SECRET
+
+"""
+# TODO(@colinschoen) Configure CI to run the full integration test suite only on protected branches like master.
+
+import os
+import random
+import unittest
+
+import requests
+
+from server.extensions import storage
+from tests.test_files import TestFile
+
+
+class EnvironmentMutationMixin(object):
+    __env_backup = {}
+
+    @classmethod
+    def set_environment_variable(cls, key, value):
+        cls.__env_backup.setdefault(key, os.environ.get(key))
+        os.environ[key] = value
+
+    @classmethod
+    def restore_environment_variable(cls, key):
+        original_value = cls.__env_backup.get(key)
+
+        if original_value is None:
+            del os.environ[key]
+        else:
+            os.environ[key] = original_value
+
+
+class CloudTestFile(TestFile, EnvironmentMutationMixin):
+    storage_provider = ""
+    key_env_name = ""
+    secret_env_name = ""
+
+    _storage_container = "okpycloudfilestest{}".format(random.randint(0, 100000))
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        storage_key = os.environ.get(cls.key_env_name)
+        storage_secret = os.environ.get(cls.secret_env_name)
+
+        if not storage_key or not storage_secret:
+            raise unittest.SkipTest("Cloud storage credentials for {} not configured".format(cls.storage_provider))
+
+        cls.set_environment_variable("STORAGE_PROVIDER", cls.storage_provider)
+        cls.set_environment_variable("STORAGE_KEY", storage_key)
+        cls.set_environment_variable("STORAGE_SECRET", storage_secret)
+        cls.set_environment_variable("STORAGE_CONTAINER", cls._storage_container)
```

Comment:
I think this will prevent users from setting a storage container to use for the integration test via an environment variable. Can we first check to see if `STORAGE_CONTAINER` is set. If not then use `cls._storage_container`.

Reply:
I believe this is by design. The way that the code is currently designed is that the touch-point to enable/disable certain integration tests are the environment variables in key_env_name and secret_env_name. Being an abstract test base, the CloudTestFile never actually executes and each cloud-specific subclass will set the storage provider environment variable that it's testing for its tests and then we revert back to the default test storage (local) for the remainder of the tests suite.

For example, if the GCP_STORAGE_KEY and GCP_STORAGE_SECRET environment variables are set, the GoogleCloudTestFile sets the storage provider to GOOGLE_STORAGE for the 14 tests it inherits from TestFile and then we revert back to local for the rest of the test suite.

```py
class GoogleCloudTestFile(CloudTestFile):
    storage_provider = "GOOGLE_STORAGE"
    key_env_name = "GCP_STORAGE_KEY"
    secret_env_name = "GCP_STORAGE_SECRET"

    test_prefix_expected_obj_name = 'test/fizz.txt'
    test_malicious_directory_traversal_expected_obj_name = 'test/_/_/fizz.txt'
```

Similarly, if we wanted to make a additional integration test for S3 storage, we'd add a new class like this:

```py
class AWSCloudTestFile(CloudTestFile):
    storage_provider = "S3"
    key_env_name = "S3_STORAGE_KEY"
    secret_env_name = "S3_STORAGE_SECRET"

    test_prefix_expected_obj_name = 'test/fizz.txt'
    test_malicious_directory_traversal_expected_obj_name = 'test/_/_/fizz.txt'
```

This setup of the code enables us to run integration tests for the 14 storage tests against multiple backends (local, GCP, AWS, etc.) during the same test run without having to re-run the entire test suite (100+ tests) for each of the storage backends.

Hope that clears things up (and I'd be happy to hear suggestions on how this could be made clearer in the code to avoid future confusion).

If we want to run the entire test suite for each of the storage providers (instead of just the TestFile cases), let me know and I can make that change.

Reply:
Maybe I am misunderstanding, but I am referring to `STORAGE_CONTAINER` not `STORAGE_PROVIDER`.

Reply:
Apologies, I completely misread that for some reason. Implemented respecting of any pre-existing storage container value in 8799fcd.